packages/ui-components/src/components/Inputs/FileUploadField.tsx

```diff
@@ -1,172 +1,291 @@
-    }
-
-    if (maxSizeInBytes) {
-      for (const file of Array.from(input.files)) {
-        const { size: newSizeInBytes } = file;
-        if (newSizeInBytes > maxSizeInBytes) {
-          setSizeInBytes(null);
-          setError(
-            `Error: file is too large: ${humanFileSize(
-              newSizeInBytes,
-            )}. Max file size: ${humanFileSize(maxSizeInBytes)}`,
-          );
-          onChange(event, undefined, null);
-          return;
-        }
-      }
-    }
-
-    if (multiple) {
-      onChange(event, newName, input.files);
-      // We don't support file size label if multiple
-    } else {
-      const [file] = Array.from(input.files);
-      setSizeInBytes(file.size);
-      onChange(event, newName, input.files);
-    }
-    setError(null);
+  if (disabled)
+    return (
+      <Uploader>
+        <SelectedFileList $doesNeedBorder={false}>
+          <SelectedFileListItem>
+            {fileName ?? <em>File uploader disabled</em>}
+            <RemoveButton disabled>Remove</RemoveButton>
+          </SelectedFileListItem>
+        </SelectedFileList>
+      </Uploader>
+    );
+
+  /**
+   * `useDropzone` can provide an `acceptedFiles` array, but it provides no way to programmatically
+   * add/remove elements. We manage file selection state manually for some custom behaviour:
+   *
+   *   1. dropping files ADDS to the selection (rather than replacing it); and
+   *   2. files may be removed individually from the selection.
+   *
+   * This array uses set semantics. See {@link onDropAccepted}.
+   */
+  const [files, setFiles] = useState<File[]>([]);
+  const hasFileSelected = files.length > 0;
+
+  /**
+   * Unions the newly selected files with the existing selection.
+   *
+   * @privateRemarks Cannot rely on `file.name` as a unique identifier, because it is only the
+   * basename. The user may select two files with the same basename from different folders.
+   * Duplicate detection in this function stringifies the whole `File` object because it is unlikely
+   * that two different files will have the same name, size AND last-modified date.
+   *
+   * @param acceptedFiles The newly selected files, provided by `react-dropzone`
+   */
+  const onDropAccepted = (acceptedFiles: File[]) => {
+    const deduped = acceptedFiles.filter(
+      acceptedFile =>
+        !files.map(file => JSON.stringify(file)).includes(JSON.stringify(acceptedFile)),
+    );
+    const newFiles = files.concat(deduped);
+    setFiles(newFiles);
   };
 
-  const removeFile = () => {
-    const input = inputEl.current;
-    if (input) {
-      input.value = '';
-      onChange(null);
-      setSizeInBytes(null);
-    }
+  const removeFile = (fileToRemove: File) => {
+    const newFiles = files.filter(file => JSON.stringify(file) !== JSON.stringify(fileToRemove));
+    setFiles(newFiles);
   };
 
+  /** Propagates file selection changes to parent */
+  useEffect(() => {
+    onChange(files);
+  }, [files]);
```

Comment:
Heya @alexd-bes, do you know how I can “prove” that `onChange` isn’t needed as a dependency here?

I tried using `[files, onChange]` to suppress the linting warning, thinking it was inconsequential because `onChange` is always the “same” function. But React hits a max. depth limit—I suspect because creates a new instance of the function for every render.

Alternatively could leave as-is. It does work, but would be nice to cover all the dependencies if it’s not much extra work